cpu/avr8_common/thread_arch.c

```diff
@@ -281,17 +285,24 @@ void thread_yield_higher(void)
 
 void avr8_exit_isr(void)
 {
-    avr8_state &= ~AVR8_STATE_FLAG_ISR;
+    /* ATxmega requires a critical section because interrupts are always
+     * enabled.
+     */
+#ifdef CPU_ATXMEGA
+    cli();
+#endif
 
-    /* Force access to avr8_state to take place */
-    __asm__ volatile ("" : : : "memory");
+    --GPIOR0;
 
-    if (sched_context_switch_request) {
-        avr8_context_save();
-        sched_run();
-        avr8_context_restore();
-        __asm__ volatile ("reti");
+#ifdef CPU_ATXMEGA
+    sei();
+#endif
+
+    if (sched_context_switch_request && GPIOR0 == 0) {
+        avr8_context_switch();
     }
+
+    __asm__ volatile ("reti");
```

Comment:
```suggestion
    if (sched_context_switch_request && (GPIOR0 == 0)) {
        avr8_context_switch();
        __asm__ volatile ("reti" : : : "memory");
    }
```

Using `reti` here to return from the function is not desirable. First, the machine code generated will still include the `ret` instruction after the `reti`, as inline assembly is not visible to the compiler. But more importantly, it will allow nested IRQs.

E.g. check this: https://godbolt.org/z/jv9j89bc7

You'll see that the call to `avr8_exit_isr()` will now return to the epilog of the ISR with IRQs enabled. It will only re-enable IRQs a few cycles before the `reti` of the ISR is reached - not buying much in terms of real time properties. But the stack can still contain all the register contents at that point in time. If e.g. in a timer IRQ a short timer is set that triggers during the ISR, it will cause a nested IRQ. If that happens a few times, the stack will overflow for little gain in real time properties.

Also: I think a memory barrier might be a good idea. It doesn't really seem to be needed with the current code, but it won't hurt to have it in place just in case.

Reply:
Understood, this answer my old question.
I'll move **reti** inside **if** and add memory barrier.
Thanks!